qml/Components/Wallpaper.qml

```diff
@@ -20,12 +21,76 @@ import Ubuntu.Components 1.3
 
 Item {
     id: root
-    property url source
+    objectName: "Wallpaper"
+    property url source: ""
+
+    // sourceSize should be set to the largest width or height that this
+    // Wallpaper will likely be set to.
+    // In most cases inside Lomiri, this should be set to the width or height
+    // of the shell, whichever is larger.
+    // However, care should be taken to avoid changing sourceSize once
+    // it's set.
+    property real sourceSize: -1.0
 
     Image {
+        id: image
+        objectName: "wallpaperImage"
         anchors.fill: parent
         fillMode: Image.PreserveAspectCrop
 
-        source: root.source
+        asynchronous: true
+
+        property real oldSourceSize
+        property real intermediarySourceSize: {
+            // If minSourceSize is defined by root's parent, it will be 0 until
+            // the binding resolves. Otherwise, it'll be -1.0. Abuse this
+            // behavior to avoid changing sourceSize more than once during
+            // startup
+            if (root.sourceSize === 0.0) {
+                developerSanityTimer.start();
+                // -1.0 signals that the wallpaper shouldn't render yet.
+                return -1.0
+            }
+            developerSanityTimer.stop();
+            if (root.sourceSize === -1.0 && root.source !== "") {
+                console.warn(`${root.objectName}'s sourceSize is unset. It will use a lot of memory.`);
+                return 0.0
+            }
+
+            return root.sourceSize
+        }
+
+        onIntermediarySourceSizeChanged: {
+            if (oldSourceSize !== 0.0 && oldSourceSize !== -1.0 && root.source !== "") {
+                console.warn(`${root.objectName} source size is changing from ${oldSourceSize} to ${intermediarySourceSize}. This will cause an expensive image reload.`);
+            }
+            oldSourceSize = intermediarySourceSize;
+        }
+
+        sourceSize: Qt.size(intermediarySourceSize, intermediarySourceSize)
```

Comment:
I've been thinking about this more. Since the wallpaper loading is now asynchronous, we could save more memory by passing sourceSize as a `size` based on the current orientation of the shell. The image would have to reload every time the user rotated their phone, but it wouldn't cause a UI thread pause.

I haven't benchmarked it to see how much memory we'd save, but in theory it could be worth it. Maybe I should wait until after we merge this though? It'd be pretty easy to do, just replacing `real sourceSize` / `real backgroundSourceSize` with `size` values instead.

Reply:
I'll try it out later. This is a large enough RAM savings to warrant doing first without waiting for more development.

Reply:
I wouldn't even try it. I'm not convinced that you'll get any savings there, and you risk making the rotation animation slower. I think this is very good the way it is :-)